Feature/landing page

henrievjen opened the pull request:
Created Landing Page and added React Routes. Clicking "Start Tutorial" will make route to tutorial, clicking Wiki temporary does nothing. Terminal gif on landing page will be possibly replaced by visual typing drone code. Also set up components folder and updated meta info.

henrievjen committed:
landing-page boilerplate

henrievjen committed:
title typing animation

henrievjen committed:
updated meta info

henrievjen committed:
responsive on smaller screens

henrievjen committed:
landing page disappears when click 'Start Tutorial'

henrievjen committed:
added react routes

translucentleaf commented on src/App.tsx:
```diff
@@ -1,13 +1,21 @@
-    </div>
+    <Router>
+      <div className="App">
+        <Landing></Landing>
```
```
<Landing />
```

ClayMav replied:
```suggestion
        <Landing />
```
Don't remove the indentation

tomdng commented on src/App.tsx:
```diff
@@ -1,13 +1,21 @@
-    </div>
+    <Router>
+      <div className="App">
+        <Landing></Landing>
```
```suggestion
<Landing />
```

tomdng replied:
React Components are usually written per my suggestion since it wouldn't make sense to nest more HTML inside components.

ClayMav commented on public/index.html:
```diff
@@ -2,19 +2,19 @@
   <head>
     <meta charset="utf-8" />
-    <link rel="shortcut icon" href="%PUBLIC_URL%/favicon.ico" />
+    <link rel="shortcut icon" href="%PUBLIC_URL%/multirotor-icon.png" />
```
ico and png are different formats, if you want to use the icon, please use some online program to convert our png into an ico, name it "favicon.ico" and place it in the correct spot

ClayMav commented on public/index.html:
```diff
@@ -2,19 +2,19 @@
     <meta
-      name="description"
-      content="Web site created using create-react-app"
+      name=""
```
```suggestion
      name="description"
```

ClayMav commented on public/index.html:
```diff
@@ -2,19 +2,19 @@
-      name="description"
-      content="Web site created using create-react-app"
+      name=""
+      content=""
```
Add some description about what this is

ClayMav commented on public/index.html:
```diff
@@ -2,19 +2,19 @@
+      content=""
     />
-    <link rel="apple-touch-icon" href="logo192.png" />
+    <link rel="apple-touch-icon" href="multirotor-icon.png" />
```
The idea here is to create multiple versions of the icon at different sizes, so I would just remove this for now if you aren't going to get each of the correctly sized assets

ClayMav commented on public/index.html:
```diff
@@ -2,19 +2,19 @@
       user's mobile device or desktop. See https://developers.google.com/web/fundamentals/web-app-manifest/
     -->
-    <link rel="manifest" href="%PUBLIC_URL%/manifest.json" />
+    <link rel="manifest" href="%PUBLIC_URL%/multirotor-icon.png" />
```
manifest.json is not an image, and is required

tomdng replied:
https://developers.google.com/web/fundamentals/web-app-manifest
A nice page explaining the purpose of the manifest.json file.

ClayMav commented on public/manifest.json:
```diff
@@ -1,19 +1,19 @@
-      "src": "logo512.png",
+      "src": "multirotor-icon.png",
       "type": "image/png",
       "sizes": "512x512"
```
With all of these, this is inherently not true, please either create versions at each of these sizes or just leave it as it was

ClayMav commented on src/App.css:
```diff
@@ -1,22 +1,3 @@
-  color: #09d3ac;
-}
+html, body, .App {
+  height: 100vh;
```
try to use 100% instead of 100vh as much as possible

ClayMav commented on src/App.tsx:
```diff
@@ -1,13 +1,21 @@
+        <Landing></Landing>
+
+        <Switch>
+          <Redirect from="/" to="home" />
```
I would not do this, just leave / and remove all reference to home

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -0,0 +1,46 @@
+import './Landing.css';
+import terminal from './landing-images/terminal.gif';
+
+const Landing: React.FC = () => {
```
Should be a return type for the function, `JSX.Element`

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -0,0 +1,46 @@
+import terminal from './landing-images/terminal.gif';
+
+const Landing: React.FC = () => {
+    const [display, setDisplay] = useState('block');
```
Should have a type for useState
`const [display, setDisplay] = useState<string>('block')`

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -0,0 +1,46 @@
+const Landing: React.FC = () => {
+    const [display, setDisplay] = useState('block');
+
+    const startTutorial = () => {
```
need types for this

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -0,0 +1,46 @@
+    }
+
+    return (
+        <div className="Landing" style={{display: display}}>
```
```suggestion
        <div className="Landing" style={{display}}>
```

ClayMav reviewed (changes requested):
I would center the content on the landing page.

The button which goes to tutorial does as expected, but if you hit back, it does not go back to the landing. You don't need to do any of the display stuff you are doing for the landing, as react router will remove it when you route away

Before we merge this, I want use to use exclusively css-in-js, as css-in-js is the standard. I suggest using the library styled-components. It won't take long and isn't super hard, obviously bootstrap is not css-in-js but i am not super worried about it but custom components should be styled in the js files

You can ignore comments I made to your use of state in Landing.tsx as I was commenting on stuff that should be removed

tomdng commented on src/components/landing/Landing.tsx:
```diff
@@ -0,0 +1,46 @@
+                    <div className="container h-100">
+                        <div className="row align-items-center h-100">
+                            <div className="col-8 mx-auto text-center">
+                                <div className="jumbotron text-center mt-5 mt-xl-0">
```
So I see CSS classes like text-center and h-100 for being used in a div and then used again in a nested div right below it. Is that actually required to keep the same styling for the page? I'm not too familiar with bootstrap again, but the use of a class in div and immediately under it does feel a little redundant at least from an outside perspective.

tomdng replied:
I've already talked to you in person about this so just resolve this comment. No need to change anything else. Still try to do CSS-in-JS with [styled components](https://www.styled-components.com/) for the custom css you have right now and keep the bootstrap.

tomdng reviewed (changes requested):
So definitely implement the changes that Clay requested. Those will make Propguard a lot more consistent and polished like a real-world webapp. 

There's not much else that I want you to change, but whenever we see each other again I'd like to talk a little about the bunch divs inside divs inside divs and so on. If you can find a way to drastically reduce the use of nested divs, that would be great.

While this works and the result looks aesthetically pleasing, if someone else ever needs to change the landing page, it will be very intimidating to them to see a giant block of HTML from the developer's perspective all with classes that may very well drastically alter the look. Good programming is a balance between good results and good code, and great programming is something that will endure the test of time and being passed around a bunch of humans.

We don't want to end up like the Mar's Rover Base Station (their desktop app) where it's a giant tangled mess of C# code with no documentation and the newcomer (my old suitemate actually) is literally trying random code just to see how one part of the code affects the other part.

Just to reiterate, I don't actually want you to change anything else other than Clay's suggestions. So in the future I'd like for us to talk about whether or not it is better to use Bootstrap or vanilla CSS/SCSS for the sake of making our codebase intuitive in the code and the UI. Fix Clay's changes and you'll be good in my eyes.

There's this motto by design legend Dieter Rams that I go by my life, whether it's web design or programming: "Good design is as little design as possible."

henrievjen committed:
removed terminal

henrievjen committed:
styled components in js

henrievjen commented:
@ClayMav @translucentleaf I fixed the issues and changed the color scheme. The React Router does not yet route to the tutorial page because it does not exist yet.

henrievjen committed:
fixed issues and changed color scheme

tomdng committed:
Merge branch 'master' into feature/landing-page

tomdng commented on src/components/landing/Landing.tsx:
```diff
@@ -1,30 +1,79 @@
                             Propguard
-                        </h1>
+                        </Title>
                         <small>A webapp that teaches people how to use Dronekit and Python to start and launch the drones</small> <br></br>
```
So visually, I think this text is a little too small. From a development standpoint, I'd prefer if you use CSS to control the font size because the <small> tag allows you to shrink fonts to sizes even tinier than what the font supports.

The reason is that while that <small> a neat feature, doing things like that in HTML (ex: using <b> for bolding fonts) may result in the browser applying font styles that don't come from the font itself. HTML will bold fonts with <b> even if the font doesn't have a bolded version.

This means there will be an inconsistency in what think you think the bolded font should look like and what the font actually is in certain cases. So while it's a _very_ minor issue, it's something to think about. I know that designers in real life can be fairly picky so things like these are very appreciated.

tomdng commented on src/components/landing/Landing.tsx:
```diff
@@ -1,30 +1,79 @@
-        </div>
+        </Landing>
     );
 }
```
For the sake of keeping code formatting close to AirBNB rules, please change everything into 2 spaces instead of 4 spaces. I think it looks nice especially since HTML looks too bloated on 4 spaces. After we merge this we should have auto-formatting tools so you don't have to worry about this at all.

tomdng reviewed (approved):
Once you get everything to 2 tabs and move away from using the <small> tag you should be good.

henrievjen committed:
increased text size and fixed tslint fixes

ClayMav commented on package.json:
```diff
@@ -8,12 +8,17 @@
     "@types/react": "16.9.3",
     "@types/react-dom": "16.9.1",
     "@types/react-router-dom": "^5.1.0",
+    "@types/styled-components": "^4.1.19",
```
types should be dev dependencies

ClayMav commented on package.json:
```diff
@@ -8,12 +8,17 @@
+    "@types/styled-components": "^4.1.19",
     "bootstrap": "^4.3.1",
+    "eslint-config-prettier": "^6.3.0",
+    "eslint-plugin-prettier": "^3.1.1",
```
eslint stuff should be dev dependencies

ClayMav commented on package.json:
```diff
@@ -8,12 +8,17 @@
     "react-scripts": "3.1.2",
-    "typescript": "3.6.3"
+    "styled-components": "^4.4.0",
+    "typescript": "^3.6.4"
```
typescript is a dev dependency

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -1,46 +1,94 @@
 
-const Landing: React.FC = () => {
-    const [display, setDisplay] = useState('block');
+const Landing: React.FC = (): JSX.Element => {
```
Should be an SFC (stateless functional component)

ClayMav commented on src/App.tsx:
```diff
@@ -1,22 +1,25 @@
+import "../node_modules/bootstrap/dist/css/bootstrap.min.css";
+import { Landing } from "./components/landing/Landing";
 
 const App: React.FC = () => {
```
Should be a SFC (stateless functional component)

ClayMav reviewed (changes requested):
Definitely add types to the styled stuff, remember, its going to be `import styled, { AnyStyledComponent } from "styled-components";` and the type is just `AnyStyledComponent`

henrievjen committed:
tslint fixes

henrievjen commented:
I pushed the fixes. I'm guessing I don't need to really add types to everything in serviceWorker.ts? (I don't really know what that file does)

tomdng committed:
Removed TSLint in favor of ESLint with TypeScript

tomdng committed:
Added Airbnb typescript config for eslint

tomdng committed:
Resolved conflicts between default typescript-eslint and airbnb rules

tomdng committed:
Overwrote indentation setting to 2 spaces

tomdng committed:
Added tsconfig.json, removed app.test.jsx, and debugged eslint errors on serviceWorker.ts

tomdng committed:
Disabling eslint for serviceWorker.js

tomdng commented:
So all the commits I did today were to move linting to just ESLint versus ESLint and TSLint. Codacy is failing the code check for indentation errors. It also isn't picking up the .eslintignore file that tells it to ignore src/serviceWorker.js. Since @ClayMav and I discussed that Codacy may just be using another ESLint configuration, merging to master may fix the issue between ESLint and Codacy.

tomdng committed:
Updated baseUrl in tsconfig.json to src/

henrievjen committed:
fixed double landing page css issue

ClayMav commented on src/App.tsx:
```diff
@@ -1,25 +1,26 @@
+  const Main: AnyStyledComponent = styled.div`
     height: 100vh;
+    overflow-y: hidden;
   `;
```
move this outside the component

ClayMav replied:
Should also probably be the HTML5 main element type

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -1,15 +1,15 @@
   `;
 
-  const Main = styled.div`
+  const Main: AnyStyledComponent = styled.div`
```
make this a styled.main

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -20,7 +20,7 @@ const Landing: React.FC = (): JSX.Element => {
   `;
 
-  const Title = styled.div`
+  const Title: AnyStyledComponent = styled.div`
```
titles should be h1

ClayMav commented on src/components/landing/Landing.tsx:
```diff
@@ -61,17 +61,17 @@ const Landing: React.FC = (): JSX.Element => {
   `;
 
-  const Subtext = styled.span`
+  const Subtext: AnyStyledComponent = styled.span`
```
choose a better html element

ClayMav reviewed (approved):
Few small changes

henrievjen committed:
added semantic elements and moved styled components outside components
